### ISSUE TITLE
[BUG] Extension doesn't work behind an autoconfigured web proxy

### ISSUE THREAD

**Issue by EnzephaloN** [neffo/bing-wallpaper-gnome-extension#22]
Hello
I'm behind a proxy (with authentication).
In Debian I set all the proxy-stuff and also "org.gnome"-proxy-settings are set.
But I get "A network error occured: 2" as message.
How can I add proxy-informations to the script?

regards
Johannes

**Comment by neffo** [comment]
Hi Johannes, we don't currently handle that case at all. We use Soup to fetch the json & images, so it shouldn't be impossible to implement.

If you are aware of a GNOME extension that has a working implementation that you'd recommend can you let me know? I don't have a way to test this, so it could be a pain to build something from scratch.

**Comment by neffo** [comment]
We use the Soup proxy resolver, which seems like it should use the GNOME settings. Can you tell me what errors you are seeing with journalctl?

**Comment by EnzephaloN** [comment]
Hello neffo
Like I sad, I had configured our proxy in GNOME. But this doesn't help. Everytime I got "network error 2". After some playing I deinstalled your Plugin, so I'm not able to give you more informations.

**Comment by neffo** [comment]
No problem, thanks for submitting the ticket though.

The error seems to be this one:
`{ SOUP_STATUS_CANT_RESOLVE_PROXY, "Cannot resolve proxy hostname" },`

That suggests the proxy autodetection is ignoring the GNOME settings. Ugh.

**Comment by kejml** [comment]
I also sit behind a corporate proxy and fetching the images doesn't work for me. but with a different error:

```
May 10 10:22:13 pc9d061 gnome-shell[3769]: BingWallpaper extension: fetching: https://www.bing.com/HPImageArchive.aspx?format=js&idx=0&n=8&mbl=1&mkt=
May 10 10:22:13 pc9d061 gnome-shell[3769]: BingWallpaper extension: Network error occured: 4
```

If I could somehow help with debugging, I'm open to doing so - I'm also a developer, but I have zero experience with Gnome extension development, so I might need some guidance.

**Comment by neffo** [comment]
This is a long standing bug, which I don't have a direct way to test. If you are able to checkout version-41 branch of this repo, you could test if this simple fix corrects it #170 . My understanding is that this proxy resolver should already be enabled by default, but that is based on documentation alone (not tested).

**Comment by kejml** [comment]
I tried uninstalling version from gnome extension page using gnome extensions app, installing version from the branch using included `install.sh`, then I logged out in Gnome and logged back in, but the result for me is, unfortunately, the same:

```
May 10 13:30:05 pc9d061 gnome-shell[1401521]: BingWallpaper extension: next check in 86400 seconds @ local time 2022-05-11 13:30 +0200
May 10 13:30:05 pc9d061 gnome-shell[1401521]: BingWallpaper extension: fetching: https://www.bing.com/HPImageArchive.aspx?format=js&idx=0&n=8&mbl=1&mkt=
May 10 13:30:05 pc9d061 gnome-shell[1401521]: BingWallpaper extension: Network error occured: 4
May 10 13:30:05 pc9d061 gnome-shell[1401521]: BingWallpaper extension: next check in 3600 seconds @ local time 2022-05-10 14:30 +0200
```

PS: In the about tab in settings, I still see version 40, but I guess, based on the diff of #170 it is correct, right?

**Comment by neffo** [comment]
Ok just pushed a new commit which will dump Soup headers, if you want to enable debug logging it should hopefully output something like this:

```
May 10 22:59:28 ubuntu-desktop-22 gnome-shell[42952]: BingWallpaper extension: Soup: GET /HPImageArchive.aspx?format=js&idx=0&n=8&mbl=1&mkt= HTTP/1.1
May 10 22:59:28 ubuntu-desktop-22 gnome-shell[42952]: BingWallpaper extension: Soup: Soup-Debug-Timestamp: 1652187568
May 10 22:59:28 ubuntu-desktop-22 gnome-shell[42952]: BingWallpaper extension: Soup: Soup-Debug: SoupSession 1 (0x55bc0bd23460), SoupMessage 1 (0x55bc0ff0e2c0), SoupSocket 1 (0x55bc0d393eb0)
May 10 22:59:28 ubuntu-desktop-22 gnome-shell[42952]: BingWallpaper extension: Soup: Host: www.bing.com
```

**Comment by kejml** [comment]
```
May 10 15:49:08 pc9d061 gnome-shell[1501388]: BingWallpaper extension: next check in 86400 seconds @ local time 2022-05-11 15:49 +0200
May 10 15:49:08 pc9d061 gnome-shell[1501388]: BingWallpaper extension: fetching: https://www.bing.com/HPImageArchive.aspx?format=js&idx=0&n=8&mbl=1&mkt=
May 10 15:49:08 pc9d061 gnome-shell[1501388]: BingWallpaper extension: Soup: HTTP/1.1 4 Could not connect: Network is unreachable
May 10 15:49:08 pc9d061 gnome-shell[1501388]: BingWallpaper extension: Soup: Soup-Debug-Timestamp: 1652190548
May 10 15:49:08 pc9d061 gnome-shell[1501388]: BingWallpaper extension: Soup: Soup-Debug: SoupMessage 0 (0x558778f5b580)
May 10 15:49:08 pc9d061 gnome-shell[1501388]: BingWallpaper extension: Network error occured: 4
May 10 15:49:08 pc9d061 gnome-shell[1501388]: BingWallpaper extension: next check in 3600 seconds @ local time 2022-05-10 16:49 +0200
```

Of course, internet connection works perfectly fine for other apps at the same moment (using that corporate proxy)

**Comment by neffo** [comment]
Doesn't seem to be using proxy at all, no.

However, if I set my proxy to 'test:8080' (not a real host) in GNOME settings I get this:

```
May 11 00:39:07 ubuntu-desktop-22 gnome-shell[42952]: BingWallpaper extension: fetching: https://www.bing.com/HPImageArchive.aspx?format=js&idx=0&n=8&mbl=1&mkt=
May 11 00:39:07 ubuntu-desktop-22 gnome-shell[42952]: BingWallpaper extension: Soup: HTTP/1.1 2 Error resolving “test”: Temporary failure in name resolution
May 11 00:39:07 ubuntu-desktop-22 gnome-shell[42952]: BingWallpaper extension: Soup: Soup-Debug-Timestamp: 1652193547
May 11 00:39:07 ubuntu-desktop-22 gnome-shell[42952]: BingWallpaper extension: Soup: Soup-Debug: SoupMessage 0 (0x55bc0ff0e4a0)
May 11 00:39:07 ubuntu-desktop-22 gnome-shell[42952]: BingWallpaper extension: Network error occured: 2
```

This suggests the proxy detection is doing something for me. I suspect your proxy is set up a bit different however. Does wget work for that link above?

**Comment by kejml** [comment]
Aha, when set proxy mode to manual, then the download works!

But for general usage, I need to use the automatic mode, so it downloads [pac file](https://en.wikipedia.org/wiki/Proxy_auto-config) from our proxy server and that obviously doesn't work with the extension

**Comment by neffo** [comment]
Right, ok so as I suspected it might be timing related. This might need some reworking, this is technical, but if there is a signal (such as dconf key changing) when the pac file is downloaded we can restart the session to handle the changes. (Or it could be that it doesn't work with pac file proxy settings, which would be annoying.)

**Comment by kejml** [comment]
Thanks, I'll test it next week, when I'm in the office again and I'll let you know

**Comment by neffo** [comment]
Any word on if this helped with this issue?

**Comment by kejml** [comment]
Oh, sorry, I completely forgot and I was in the office yesterday.... I'll put a reminder to my calendar

**Comment by neffo** [comment]
Ok I've been testing this locally and for some reason libsoup (or at least the way I'm using it) doesn't acknowledge auto configured proxies (even with changes I've made). It's not timing related (as I suspected above), so nothing will have changed. It's like the proxy resolver is not picking up any proxy at all, even with a very simple .PAC file.

Using manual proxy works just fine (so confirming what you said).

**Comment by neffo** [comment]
Thanks for confirming (means my testing approach was useful), I've actually had a lot of trouble tracking this one down. I'm not sure libsoup actually knows what to do with .pac files. It seems to use a simple proxy lookup that appears to just look at protocols (aka the http/https/socks part of the url). I might raise a ticket upstream. An alternative, hacky method would be to manually set a proxy host within the extension but that doesn't seem the 'right' way to do it.

**Comment by kejml** [comment]
Personally, I would raise the ticket in the lib. This ticket is open for more than 4 years without getting too much attention, so I don't think it's worth implementing some hacky solution on your side.

**Comment by kejml** [comment]
@neffo Did you close this one on purpose? I guess you'll just wait for a fix in the library, right?

**Comment by neffo** [comment]
Yep, was closed automatically. Reopened, pending a miracle.

**Comment by digeomel** [comment]
Hello, still having this issue behind a corporate proxy with a .pac file. Is this being tracked in an issue with the Soup library? Maybe we  can link it here?
Thanks

**Comment by neffo** [comment]
We are using a newer soup version in extension, so there may be better handling of the .pac proxy handling.

We shouldn't need to be doing anything particularly special, so it may be we aren't setting up connections in a way that handles proxies negotiations.

Documentation was a bit lacking last time I looked so I couldn't make much progress.
